x-pack/plugins/fleet/server/services/output.ts

```diff
@@ -980,6 +986,48 @@ class OutputService {
       }
     }
 
+    // Store secret values if enabled; if not, store plain text values
+    if (await isOutputSecretStorageEnabled(esClient, soClient)) {
+      const secretsRes = await extractAndUpdateOutputSecrets({
+        oldOutput: originalOutput,
+        outputUpdate: data,
+        esClient,
+        secretHashes: data.is_preconfigured ? secretHashes : undefined,
+      });
+
+      updateData.secrets = secretsRes.outputUpdate.secrets;
+      secretsToDelete = secretsRes.secretsToDelete;
+    } else {
+      if (
+        originalOutput.type === outputType.Logstash &&
+        data.type === outputType.Logstash &&
+        updateData.type === outputType.Logstash
+      ) {
+        if (!data.ssl?.key && data.secrets?.ssl?.key) {
+          updateData.ssl = JSON.stringify({ ...data.ssl, ...data.secrets.ssl });
+        }
+      } else if (
+        originalOutput.type === outputType.Kafka &&
+        data.type === outputType.Kafka &&
+        updateData.type === outputType.Kafka
+      ) {
+        if (!data.password && data.secrets?.password) {
+          updateData.password = data.secrets?.password as string;
+        }
+        if (!data.ssl?.key && data.secrets?.ssl?.key) {
+          updateData.ssl = JSON.stringify({ ...data.ssl, ...data.secrets.ssl });
+        }
+      } else if (
+        originalOutput.type === outputType.RemoteElasticsearch &&
+        data.type === outputType.RemoteElasticsearch &&
+        updateData.type === outputType.RemoteElasticsearch
+      ) {
+        if (!data.service_token && data.secrets?.service_token) {
+          updateData.service_token = data.secrets?.service_token as string;
```

Comment:
I wonder if this logic can be extracted, seems very similar to the create method. It could be done in a follow up pr.
Do we have test coverage over this logic?

Also, is it necessary for `originalOutput` to have the same type as the updated output? It can happen that the type changes, isn't it?

Reply:
> I wonder if this logic can be extracted, seems very similar to the create method. It could be done in a follow up pr.

Agreed. These methods are very long and could use some extracting/refactoring. It's always a little dangerous to leave refactoring for followup, but it might be the right call in this case.

> Do we have test coverage over this logic?

We do in the integration tests, I suppose, but absolutely happy to add some unit tests as well.

> Also, is it necessary for originalOutput to have the same type as the updated output? It can happen that the type changes, isn't it?

Good point! I guess that is not required (and perhaps even a liability). Perhaps it would be good to add a test with an output type change.

Reply:
I think we have some tests on output type change in cypress tests.